Verify cocoon and autosubmit provenance

drewroengoogle opened the pull request:
* Adds additional steps to the deploy cloud build for autosubmit and cocoon
  * First, download the provenance from artifact registry
  * Next, use slsa-verifier to ensure the provenance fits the policy defined within `verify-provenance.sh`
  * If any of these steps fail, the build is marked red and does not proceed with deployment.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

godofredoc commented on cloud_build/verify-provenance.sh:
```diff
@@ -0,0 +1,34 @@
+# Download the jq binary in order to obtain the artifact registry url from the
+# docker image provenance.
+echo "Installing jq using curl..."
+curl -Lo jq \
```
Is there any other way to install jq? piping something from the internet to bash is considered a security bad practice.

drewroengoogle replied:
That's fair. Perhaps there is already an image out there with jq installed where it wouldn't need to be downloaded at all? I'll look for an alternative.

drewroengoogle replied:
I've updated the installation to be done through `apt`, and specified a version tag as well, since there isn't a hash option for apt.

godofredoc commented on cloud_build/verify-provenance.sh:
```diff
@@ -0,0 +1,34 @@
+
+# Download slsa-verifier in order to validate the docker image provenance.
+echo "Installing slsa-verifier using go..."
+go install github.com/slsa-framework/slsa-verifier/v2/cli/slsa-verifier@v2.0.1
```
Is it possible to use a full hash rather than a version?

drewroengoogle replied:
Updated to hash!

godofredoc commented:
@mihaimaruseac this is the PR validating the cloudbuild provenance.

godofredoc commented on cloud_build/verify-provenance.sh:
```diff
@@ -0,0 +1,32 @@
+# Copyright 2023 The Flutter Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
```
Can we add some notes about the expected exit codes of this script? e.g. this script will exit with a non zero exit code if the policy validation fails,

mihaimaruseac commented on app_dart/cloudbuild_app_dart_deploy.yaml:
```diff
@@ -5,6 +5,27 @@
+    args:
+      - '-c'
+      - |-
+        bash cloud_build/verify-slsa-provenance.sh
```
```suggestion
        bash cloud_build/verify-provenance.sh
```

mihaimaruseac commented on auto_submit/cloudbuild_auto_submit_deploy.yaml:
```diff
@@ -5,6 +5,26 @@
+    args:
+      - '-c'
+      - |-
+        bash cloud_build/verify-slsa-provenance.sh
```
```suggestion
        bash cloud_build/verify-provenance.sh
```

mihaimaruseac commented on cloud_build/verify-provenance.sh:
```diff
@@ -0,0 +1,32 @@
+  --builder-id=https://cloudbuild.googleapis.com/GoogleHostedWorker@v0.3 \
+  --provenance-path unverified-provenance.json
+
+echo "Provenance has been successfully validated!"
```
I think you don't need this additional echo, the `slsa-verifier` command prints a successful message when it validates (`PASSED: Verified SLSA provenance`) and a failure message otherwise (e.g., `FAILED: SLSA verification failed: open fibs: no such file or directory` or `FAILED: SLSA verification failed: expected hash...`)

drewroengoogle replied:
Good catch! I added that initially to my testing script and forgot to remove it.

mihaimaruseac reviewed (commented):
In principle LGTM, though the script name does not match the call.

Also, unsure if Dependabot (or similar bots) are able to update the pinned deps in the script. But that's tangential

godofredoc commented:
@drewroengoogle any updates on this PR?

drewroengoogle commented:
> @drewroengoogle any updates on this PR?

Yes, I got distracted with some other problems, but switching over to this now. I've updated the PR addressing comments, so feel free to approve again if you feel we can merge it

mihaimaruseac commented on app_dart/cloudbuild_app_dart_deploy.yaml:
```diff
@@ -5,6 +5,27 @@
+    args:
+      - '-c'
+      - |-
+        bash cloud_build/verify-provenance.sh
```
Very minor nit: I'm not sure if this could be simplified to just `cloud_build/verify-provenance.sh` (so that what gets run is `/bin/bash -c cloud_build/verify-provenance.sh` instead of `/bin/bash -c bash cloud_build/verify-provenance.sh`.

drewroengoogle replied:
Good point, this was mostly following similar cloud build scripts we had, but after trying it out, it will still run correctly. Updated!

mihaimaruseac commented on cloud_build/verify-provenance.sh:
```diff
@@ -0,0 +1,33 @@
+# the provenance metadata.
+echo "Verifying the provenance is valid and correct..."
+slsa-verifier verify-image $(cat unverified-provenance.json | \
+  ./jq -r .image_summary.fully_qualified_digest) \
```
Minor nit: I would extract the `$(cat unverified-provenance.json | ./jq -r .image_summary.fully_qualified_digest)` part to a separate variable to make this slightly more readable. On first glance it looked like the `slsa-verifier` arguments following this line were being passed to `./jq` instead (shouldn't this be `jq` instead?)

drewroengoogle replied:
Updated to move the digest part to its own variable. I also changed it to `jq` - that was left over from when we switched to installing jq from apt instead of directly getting the binary.

mihaimaruseac reviewed (commented):
LGTM, just nits

mihaimaruseac reviewed (commented):
LGTM
